src/tqec/templates/display.py

```diff
@@ -29,6 +30,88 @@ def display_template(
         print()
 
 
+def display_templates_ascii(
+    templates: ComposedTemplate,
+    plaquette_indices: ty.Sequence[int] | None = None,
+    h_space: int = 5,
+    v_space: int = 2,
+    corner_mark: str = "■",
+) -> None:
+    """Display the templates as an ASCII art (multi-line string).
+
+    Args:
+        templates: the ComposedTemplate instance to display.
+        plaquette_indices: the plaquette indices that are forwarded to the call
+            to `template.instantiate` to get the actual template representation.
+        h_space: horizontal space allocated to each plaquette location (for index and separation element)
+        v_space: vertical space between two rows of plaquettes (value = 1 + num additional lines)
+    """
+    if plaquette_indices is None:
+        plaquette_indices = tuple(range(1, templates.expected_plaquettes_number + 1))
+    if h_space < 4:
+        print(f"WARNING: horizontal space value {h_space} is not enough. Changed to default value 5.")
+        h_space = 5
+    if v_space < 2:
+        print(f"WARNING: vertical space value {v_space} is not enough. Changed to default value 2.")
+        v_space = 2
+    if len(corner_mark) != 1:
+        print(f"WARNING: corner mark '{corner_mark}' is unacceptable. Changed to default '■'")
+        corner_mark = "■"
+    # Numpy array including all templates.
+    arr = templates.instantiate(plaquette_indices)
+    # Upper-left and bottom-right corners of every template, expressed as (row, column) of the array returned.
+    ul_pos = {}
+    br_pos = {}
+    ul_positions = templates._compute_ul_absolute_position()
+    bbul, _ = templates._get_bounding_box_from_ul_positions(ul_positions)
+    for tid, tul in ul_positions.items():
+        # We follow the same procedure as in Template.instantiate():
+        # Subtracting bbul (upper-left bounding box position) from each coordinate to stick
+        # the represented code to the axes and avoid having negative indices.
+        x = tul.x - bbul.x
+        y = tul.y - bbul.y
+        # Recall that numpy indexing is (y, x) in our coordinate system convention.
+        tshapey, tshapex = templates._templates[tid].shape.to_numpy_shape()
+        ul_pos[tid] = (y, x)
+        br_pos[tid] = (y + tshapey, x + tshapex)
+    # Format of the ASCII art.   
+    x_size = numpy.shape(arr)[1]
+    empty_line = " "*x_size*h_space
+    # Instead of printing to screen, we create a multi-line buffer string.
+    # When completed, we print the buffer.
+    buffer_lines = []
+    buffer_lines.append(empty_line)
+    for line in arr:
+        buffer = ""
+        for element in line:
+            element = str(element) if element != 0 else "."
+            buffer += f" {element:^{h_space-1}}"
+        buffer_lines.append(buffer)
+        for _ in range(v_space-1):
+            buffer_lines.append(empty_line)
```

Comment:
It would be a little bit more readable to have vertically-centered plaquette numbers (they are already correctly horizontally centered).

I *think* the following should work:
```suggestion
    buffer_lines = []
    buffer_lines.append(empty_line)
    empty_lines_number_before = (v_space - 1) // 2
    empty_lines_number_after = v_space - 1 - empty_lines_number_before
    for line in arr:
        for _ in range(empty_lines_number_before):
            buffer_lines.append(empty_line)
        buffer = ""
        for element in line:
            element = str(element) if element != 0 else "."
            buffer += f" {element:^{h_space-1}}"
        buffer_lines.append(buffer)
        for _ in range(empty_lines_number_after):
            buffer_lines.append(empty_line)
```
from my local tests, that works but the labels are one line too low on odd `v_space` values (personal taste).